File: Ant/chapter3/python/jack_straws.py
EPS = 1e-10


def ns(f):
    return next(f).strip()


def add(a, b):
    if abs(a + b) < EPS * (abs(a) + abs(b)):
        return 0.0
    return float(a + b)


class P:
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def __add__(self, other):
        return P(add(self.x, other.x), add(self.y, other.y))

    def __sub__(self, other):
        return P(add(self.x, -other.x), add(self.y, -other.y))

    def __mul__(self, d):
        return P(self.x * d, self.y * d)

    def dot(self, other):
        return add(self.x * other.x, self.y * other.y)

    def det(self, other):
        return add(self.x * other.y, - self.y * other.x)


def on_seg(p1, p2, q):
    return (p1 - q).det(p2 - q) == 0 and (p1 - q).dot(p2 - q) <= 0


def intersection(p1, p2, q1, q2):
    return p1 + (p2 - p1) * ((q2 - q1).det(q1 - p1) / (q2 - q1).det(p2 - p1))


with open("../testset/jack_straws/test1.txt", 'r') as f:
    N, M = map(int, ns(f).split())
    p = [P(*map(int, ns(f).split())) for _ in range(N)]
    q = [P(*map(int, ns(f).split())) for _ in range(N)]
    c = [list(map(lambda x: int(x) - 1, ns(f).split())) for _ in range(M)]


g = [[False] * N for _ in range(N)]
for i in range(N):
    g[i][i] = True
    for j in range(i):
        if (p[i] - q[i]).det(p[j] - q[j]) == 0:
            # parallel
            g[i][j] = g[j][i] = on_seg(p[i], q[i], p[j]) or on_seg(
                p[i], q[i], q[j]) or on_seg(p[j], q[j], p[i]) or on_seg(p[j], q[j], q[i])
        else:
            # not parallel
            r = intersection(p[i], q[i], p[j], q[j])
            g[i][j] = g[j][i] = on_seg(p[i], q[i], r) and on_seg(p[j], q[j], r)
for k in range(N):
    for i in range(N):
        for j in range(N):
            g[i][j] |= g[i][k] and g[k][j]
for a, b in c:
    if g[a][b]:
        print("CONNECTED")
    else:
        print("NOT CONNECTED")

File: Ant/chapter3/cpp/jack_straws.cpp
#include <fstream>
#include <stdio.h>

#define MAX_N 12
#define MAX_M 10000

using namespace std;

double EPS = 1e-10;

double add(double a, double b) {
  if (abs(a + b) < EPS * (abs(a) + abs(b))) {
    return 0;
  }
  return a + b;
}

struct P {
  double x, y;
  P() {}
  P(double x, double y) : x(x), y(y) {}
  P operator+(P p) { return P(add(x, p.x), add(y, p.y)); }
  P operator-(P p) { return P(add(x, -p.x), add(y, -p.y)); }
  P operator*(double d) { return P(x * d, y * d); }
  double dot(P p) { return add(x * p.x, y * p.y); }
  double det(P p) { return add(x * p.y, -y * p.x); }
};

bool on_seg(P p1, P p2, P q) { return (p1 - q).det(p2 - q) == 0 && (p1 - q).dot(p2 - q) <= 0; }

P intersection(P p1, P p2, P q1, P q2) {
  return p1 + (p2 - p1) * ((q2 - q1).det(q1 - p1) / (q2 - q1).det(p2 - p1));
}

int N, M;
P p[MAX_N], q[MAX_N];
int a[MAX_M], b[MAX_M];
bool g[MAX_N][MAX_N];

void solve() {
  for (int i = 0; i < N; ++i) {
    g[i][i] = true;
    for (int j = 0; j < i; ++j) {
      if ((p[i] - q[i]).det(p[j] - q[j]) == 0) {
        // parallel
        g[i][j] = g[j][i] = on_seg(p[i], q[i], p[j]) || on_seg(p[i], q[i], q[j]) ||
                            on_seg(p[j], q[j], p[i]) || on_seg(p[j], q[j], q[i]);
      } else {
        // not parallel
        P r = intersection(p[i], q[i], p[j], q[j]);
        g[i][j] = g[j][i] = on_seg(p[i], q[i], r) && on_seg(p[j], q[j], r);
      }
    }
  }
  // Warshall-Floyd algorithm
  for (int k = 0; k < N; ++k) {
    for (int i = 0; i < N; ++i) {
      for (int j = 0; j < N; ++j) {
        g[i][j] |= g[i][k] && g[k][j];
      }
    }
  }
  for (int i = 0; i < M; ++i) {
    printf("%s\n", g[a[i]][b[i]] ? "CONNECTED" : "NOT CONNECTED");
  }
}

int main() {
  ifstream ifs("../testset/jack_straws/test1.txt");
  ifs >> N >> M;
  for (int i = 0; i < N; ++i) {
    int p1, p2;
    ifs >> p1 >> p2;
    p[i] = P(p1, p2);
  }
  for (int i = 0; i < N; ++i) {
    int q1, q2;
    ifs >> q1 >> q2;
    q[i] = P(q1, q2);
  }
  for (int i = 0; i < M; ++i) {
    ifs >> a[i] >> b[i];
    --a[i];
    --b[i];
  }
  solve();
}